boot/boot.go

```diff
@@ -429,8 +429,50 @@ func updateManagedBootConfigForBootloader(dev Device, mode, gadgetSnapOrDir stri
 		return false, err
 	}
 	// boot config update can lead to a change of kernel command line
-	if err := observeCommandLineUpdate(dev.Model(), gadgetSnapOrDir); err != nil {
+	_, err = observeCommandLineUpdate(dev.Model(), commandLineUpdateReasonSnapd, gadgetSnapOrDir)
+	if err != nil {
 		return false, err
 	}
 	return tbl.UpdateBootConfig()
 }
+
+// UpdateCommandLineForGadgetComponent handles the update of a gadget that
+// contributes to the kernel command line of the run system. Returns true when a
+// change in command line has been observed and a reboot is needed. The reboot,
+// if needed, should be requested as the the earliest possible occasion.
+func UpdateCommandLineForGadgetComponent(dev Device, gadgetSnapOrDir string) (needsReboot bool, err error) {
+	if !dev.HasModeenv() {
+		// only UC20 devices are supported
+		return false, fmt.Errorf("internal error: command line component cannot be updated on non UC20 devices")
+	}
+	opts := &bootloader.Options{
+		Role:        bootloader.RoleRunMode,
+		NoSlashBoot: true,
+	}
+	tbl, err := getBootloaderManagingItsAssets(InitramfsUbuntuBootDir, opts)
+	if err != nil {
+		if err == errBootConfigNotManaged {
+			// we're not managing this bootloader's boot config
```

Comment:
the pi has a cmdline.txt already? how would this interact with that? to be clear I don't think we can make this a generalized feature, we might expand the set of bootloaders supporting it but we need a clear use case, because general availability in itself is probably not achievable

Reply:
The pi's cmdline.txt only really comes into play for us if we were booting the kernel directly from the Pi's GPU bootloader; since we use u-boot in the middle cmdline.txt doesn't matter and is effectively ignored by u-boot AIUI (though this is probably different in classic world, I am only talking about Ubuntu Core). We may someday only use the Pi's GPU bootloader directly in which case this feature would instead mutate cmdline.txt I think instead of writing to the bootenv var, but again that is in the future after we drop u-boot (which AIUI is not actively worked on by anyone and is still a bit of a stretch goal for someday).

I think the use case is what I mentioned above: making it easy to customize the kernel command line without needing to rebuild binary artifacts in the gadget snaps that Canonical publishes (which you need to do today). We can modify the u-boot boot script in our Pi gadget to support reading this bootenv variable (or the full bootenv var too) and add/replace the kernel command line from there too. 

Also I don't really see the harm in enabling the writing of this bootenv for bootloaders that may not support it, nothing will be broken by writing an additional bootenv variable that the bootloader ignores.

Reply:
> Also I don't really see the harm in enabling the writing of this bootenv for bootloaders that may not support it, nothing will be broken by writing an additional bootenv variable that the bootloader ignores.

that POV is completely ignoring the expectation of the user. We can't tell you can set cmdline.extra but it might not work sometimes, just try.